### PR TITLE
gnrc_netif_lorawan: add support for LINK_UP/_DOWN events

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This prepares for SCHC support with LoRaWAN, by adding UP/DOWN events to `gnrc_netif_lorawan`. Since with LoRaWAN there is only a hardware address (and thus an IID) after the device joined the network (which is the UP event), auto-configuration of link-local addresses needs to be done on UP. Since the hardware address (and the IID for that matter, see https://datatracker.ietf.org/doc/html/rfc9011#section-5.3) might change on the next connect, the link-local address needs to be removed on down.

### Testing procedure
Sadly, there is no IPv6 support for LoRaWAN yet, so only a compile test must suffice for now. However, I need this for SCHC and IPv6 support (i.e. SCHC) for LoRaWAN is needed to test this, so this is somewhat of a hen-and-egg problem...

### Issues/PRs references
In preparation for #18700.

Requires ~~#18708~~ (merged).

### CODE REVIEW

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c [hunk 1]
```diff
@@ -574,10 +591,13 @@ static int _set(gnrc_netif_t *netif, const gnrc_netapi_opt_t *opt)
         }
         else {
             mlme_request.type = MLME_RESET;
+
```
Comment (jia200x) [inline sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:594]: unrelated change

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c [hunk 2]
```diff
@@ -69,7 +70,14 @@ void gnrc_lorawan_mlme_confirm(gnrc_lorawan_t *mac, mlme_confirm_t *confirm)
 
     if (confirm->type == MLME_JOIN) {
         if (confirm->status == 0) {
-            DEBUG("gnrc_lorawan: join succeeded\n");
+            gnrc_netif_lorawan_t *lw_netif = container_of(mac, gnrc_netif_lorawan_t, mac);
+            gnrc_netif_t *netif = container_of(lw_netif, gnrc_netif_t, lorawan);
+
+            printf("gnrc_lorawan: join succeeded %d\n", netif->pid);
+            if (netif->dev) {
+                netif->lorawan.flags |= GNRC_NETIF_LORAWAN_FLAGS_JOINED;
```
Comment (jia200x) [inline sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:78]: this flag seems to be write only. Do you plan to use it later?

Comment (miri64) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:78]: Yes, for the IID getter I plan to use it: If the interface is connected, return the IID, but -ENOTCONN otherwise. This way the error check introduced in this PR for getting the IID in the NIB, will also guarantee, that the device does not get an address for an empty IID. Using a flag is far less complicated than checking the UP status via MLME (as the structs require nearly 100 bytes on the stack if used).

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c (hunk 1 above)
Comment (miri64) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c]: Addressed.

**Comment by miri64** [comment]
Rebased and adopted for the change introduced in https://github.com/RIOT-OS/RIOT/pull/18708

**Comment by miri64** [comment]
Rebased to current master. No longer waiting for another PR.

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c [hunk 3]
```diff
@@ -136,6 +136,7 @@ void gnrc_ipv6_nib_iface_up(gnrc_netif_t *netif)
     netif->ipv6.rs_sent = 0;
 #endif  /* CONFIG_GNRC_IPV6_NIB_6LN */
     netif->ipv6.na_sent = 0;
+    _auto_configure_addr(netif, &ipv6_addr_link_local_prefix, 64U);
```
Comment (benpicco) [inline sys/net/gnrc/network_layer/ipv6/nib/nib.c:139]: If we do this in `gnrc_ipv6_nib_iface_up()` we can remove it from `gnrc_ipv6_nib_init_iface()`

Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c:139]: Done, but not tested so far.

**Review by benpicco — APPROVED** [review]
`examples/gnrc_border_router` still works on `esp32-wroom-32`

```
2022-10-11 22:52:47,163 # Iface  9  HWaddr: 3C:71:BF:9E:13:FD  Channel: 6 
2022-10-11 22:52:47,166 #           L2-PDU:249  MTU:1280  HL:64  RTR  
2022-10-11 22:52:47,170 #           6LO  Source address length: 6
2022-10-11 22:52:47,172 #           Link type: wireless
2022-10-11 22:52:47,178 #           inet6 addr: fe80::3e71:bfff:fe9e:13fd  scope: link  VAL
2022-10-11 22:52:47,181 #           inet6 group: ff02::2
2022-10-11 22:52:47,184 #           inet6 group: ff02::1
2022-10-11 22:52:47,187 #           inet6 group: ff02::1:ff9e:13fd
2022-10-11 22:52:47,188 #           
2022-10-11 22:52:47,193 # Iface  10  HWaddr: 3C:71:BF:9E:13:FC  Channel: 0  Link: down 
2022-10-11 22:52:47,197 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2022-10-11 22:52:47,200 #           Source address length: 6
2022-10-11 22:52:47,203 #           Link type: wireless
2022-10-11 22:52:47,206 #           inet6 group: ff02::2
2022-10-11 22:52:47,208 #           inet6 group: ff02::1
2022-10-11 22:52:47,209 #           

2022-10-11 22:52:49,254 # WiFi connected to ssid beersperminute, channel 1

2022-10-11 22:52:52,456 # > ifconfig
2022-10-11 22:52:52,461 # Iface  9  HWaddr: 3C:71:BF:9E:13:FD  Channel: 1 
2022-10-11 22:52:52,464 #           L2-PDU:249  MTU:1280  HL:64  RTR  
2022-10-11 22:52:52,468 #           6LO  Source address length: 6
2022-10-11 22:52:52,470 #           Link type: wireless
2022-10-11 22:52:52,476 #           inet6 addr: fe80::3e71:bfff:fe9e:13fd  scope: link  VAL
2022-10-11 22:52:52,479 #           inet6 group: ff02::2
2022-10-11 22:52:52,482 #           inet6 group: ff02::1
2022-10-11 22:52:52,485 #           inet6 group: ff02::1:ff9e:13fd
2022-10-11 22:52:52,486 #           
2022-10-11 22:52:52,491 # Iface  10  HWaddr: 3C:71:BF:9E:13:FC  Channel: 1  Link: up 
2022-10-11 22:52:52,496 #           L2-PDU:1500  MTU:1492  HL:255  RTR  
2022-10-11 22:52:52,498 #           Source address length: 6
2022-10-11 22:52:52,501 #           Link type: wireless
2022-10-11 22:52:52,507 #           inet6 addr: fe80::3e71:bfff:fe9e:13fc  scope: link  VAL
2022-10-11 22:52:52,514 #           inet6 addr: 2001:9e8:1436:e500:3e71:bfff:fe9e:13fc  scope: global  VAL
2022-10-11 22:52:52,517 #           inet6 group: ff02::2
2022-10-11 22:52:52,519 #           inet6 group: ff02::1
2022-10-11 22:52:52,522 #           inet6 group: ff02::1:ff9e:13fc
2022-10-11 22:52:52,524 #   
```

Please squash

**Comment by miri64** [comment]
> 2022-10-11 22:52:49,254 # WiFi connected to ssid beersperminute, channel 1

I see the WG SSID did not change in the past 10 years :D

**Comment by miri64** [comment]
Iface 9 is not WiFi, I assume?

**Comment by miri64** [comment]
(because it says "wireless")

**Comment by benpicco** [comment]
Iface 9 is `esp_now`.
All ESPs can run both in parallel and act as a WiFi <-> esp_now border router.

**Comment by miri64** [comment]
Mh, the test expects the address to be configured. Wasn't there some way to wait for the up event?

**Comment by miri64** [comment]
Ah, since it is only a mock interface, I can just call `_common_set_up()` which handles the bootstrapping of the interface between tests sufficiently.
